models/offer.go

```diff
@@ -337,3 +353,19 @@ func filterTemplatesByFrequencyAndPeriod(offers []*Offer, ots []*OfferTemplate,
 
 	return filteredOts, nil
 }
+
+func filterTemplatesByClaimedOffers(offersByKey map[string]*Offer, ots []*OfferTemplate) ([]*OfferTemplate, error) {
```

Comment:
Here you are only allowing an offer to be claimed once, right? It would be better to check the number of purchases against the max period. Also, when you fix the issue above and have several offers for the same offer template key you need to aggregate the number of purchases before comparing.